### PR TITLE
Fully split ReactDOM from React

### PR DESCRIPTION
**Pull request by iamdustan**
Based on this thread: https://twitter.com/sebmarkbage/status/692115760592871424

This is definitely not ready for shipping, but appears to work from first glance.

With this change:

* `npm test` still passes
* `npm run build && open examples/basic-click-counter/index.html` works as expected
* `grunt compare-size:files`:

```
Running "compare_size:files" (compare_size) task
   raw     gz Sizes
   793    490 build/react-dom-server.js
   736    446 build/react-dom-server.min.js
131713  37977 build/react-dom.js
   717    437 build/react-dom.min.js
688454 155471 build/react-with-addons.js
147913  42915 build/react-with-addons.min.js
104904  25095 build/react.js
 21106   7102 build/react.min.js

   raw     gz Compared to inject-owner @
              f05ea8fd47261d4877e1f54b6d140a22b2af1a6b
  -408   -153 build/react-dom-server.js
     =      = build/react-dom-server.min.js
+130531+37343 build/react-dom.js
     =      = build/react-dom.min.js
     =      = build/react-with-addons.js
     =      = build/react-with-addons.min.js
     =      = build/react.js
     =      = build/react.min.js

   raw     gz Compared to last run
  -408   -153 build/react-dom-server.js
     =      = build/react-dom-server.min.js
+130531+37343 build/react-dom.js
     =      = build/react-dom.min.js
     =      = build/react-with-addons.js
     =      = build/react-with-addons.min.js
     =      = build/react.js
     =      = build/react.min.js
```

Haven’t had a chance to dig in super deep yet. Will look more in-depth later this week unless this is just a terrible idea as is. :smile:

### CODE REVIEW

File: grunt/tasks/react-dom.js [hunk 1]
```diff
@@ -25,6 +25,8 @@ function build(name, filename) {
 }
 
 module.exports = function() {
+  /*
   build('ReactDOM', 'react-dom');
   build('ReactDOMServer', 'react-dom-server');
+  */
```
Comment (iamdustan) [inline grunt/tasks/react-dom.js]: this file should probably be deleted now that it’s built from source and not copying these over.

File: grunt/config/browserify.js [hunk 2]
```diff
@@ -109,9 +109,46 @@ var addonsMin = {
   after: [minify, bannerify],
 };
 
+var reactDom = {
```
Comment (iamdustan) [inline grunt/config/browserify.js]: TODO: these appeared to only build minified versions. Need to generate both regular and .min builds

File: src/isomorphic/classic/element/ReactCurrentOwner.js [hunk 3]
```diff
@@ -18,13 +18,11 @@
...
 var ReactCurrentOwner = {
-
-  /**
-   * @internal
-   * @type {ReactComponent}
-   */
   current: null,
 
+  inject: function(InjectedCurrentOwner) {
+    ReactCurrentOwner = InjectedCurrentOwner;
```
Comment (iamdustan) [inline src/isomorphic/classic/element/ReactCurrentOwner.js]: The `InjectedCurrentOwner` is just `{current: null}` which works by reference, but does feel slightly hackish.

**Comment by sebmarkbage** [comment]
We don't want any dynamic injection but luckily we don't need to because there is only one implementation of CurrentOwner. You just need to make sure that the module instance is shared by all Reacts.

To do that I think that we probably want to break it out into a separate package on npm with a frozen version. That way it works with deduped packages even if there are multiple versions of the main React package.

**Comment by sebmarkbage** [comment]
It doesn't look like this actually has a way to share the ReactCurrentOwner instance between the three packages. There is no coordination mechanism. You simply end up with three different ReactCurrentOwner if you use the react, react-dom, and react-dom-server packages on the same page.

This will break string refs and debugging messages.

**Comment by iamdustan** [comment]
@sebmarkbage so release a `ReactCurrentOwner@1.0.0` which React, ReactDOM, ReactNative, React* would all be required to depend on? Is there any reason to put that in `fbjs` or keep stand alone?

**Comment by sebmarkbage** [comment]
Yea. It shouldn't go into fbjs. This is react-only. It can't go into the fbjs package anyway since it is versioned differently.

**Comment by iamdustan** [comment]
> You simply end up with three different ReactCurrentOwner if you use the react, react-dom, and react-dom-server packages on the same page.

Good point. I just went through the rest of the demos and a few did not work, but of course I only generated minified builds so didn’t get real debugging messages. I was just about to checkout master to see if I broke them or if they were already in that state.

I was thinking that `react`’s would be overwritten as soon as `react-dom` or `react-dom-server` called their own `DefaultInjection`, though since they overwrite react using those in tandem would break it. (Yes, I realize I’m repeating what you just said. Just making sure I’m following correctly.)

**Comment by iamdustan** [comment]
Does this mean browser builds would need to load `react.js`, `react-current-owner.js`, `react-dom.js`, and optionally `react-dom-server.js`? Or would ReactCurrentOwner be shared in a `React.__SECRET_CURRENT_OWNER` property?

**Comment by jimfb** [comment]
PR closed in favor of discussion in https://github.com/facebook/react/issues/5974

**Pull request closed without merging**